Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php

```diff
@@ -250,7 +282,45 @@ public function retrieveNodePath(NodeAggregateId $nodeAggregateId): NodePath
         if ($result === []) {
             throw new \InvalidArgumentException(sprintf('Failed to retrieve node path for node "%s"', $nodeAggregateId->value), 1678391715);
         }
-        return NodePath::fromPathSegments(array_reverse(array_column($result, 'name')));
+        $rows = array_reverse($result);
+        $rootNodeTypeNameQueryBuilder = $this->createQueryBuilder()
+            ->select('nodetypename')
+            ->from($this->tableNamePrefix . '_node')
+            ->where('relationanchorpoint = :relationAnchorPoint')
+            ->setParameter('relationAnchorPoint', $rows[0]['childnodeanchor']);
+        $rootNodeTypeNameRow = $this->executeQuery($rootNodeTypeNameQueryBuilder)->fetchAssociative();
+        if ($rootNodeTypeNameRow === false) {
+            throw new \InvalidArgumentException(sprintf('Failed to retrieve root node type for node "%s"', $nodeAggregateId->value), 1687111057);
+        }
+        $rootNodeTypeName = NodeTypeName::fromString($rootNodeTypeNameRow['nodetypename']);
+
+        $pathSegments = [];
+        foreach ($rows as $i => $row) {
+            if ($i !== 0) { // skip the root node, it's resolved by node type name
+                if ($row['name']) {
```

Comment:
@nezaniel i would find it actually cool to use findAncestors https://github.com/neos/neos-development-collection/pull/4349 instead and just postprocess the nodes.
This would would reduce complexity quite a bit ^^

Reply:
Wow, this is the most complex code in this class now and it exceeds my mental capabilities

> This would would reduce complexity quite a bit ^^

Yes, please :)

Reply:
We said, we want to finish https://github.com/neos/neos-development-collection/pull/4349 first ^^